main.py

```diff
@@ -345,31 +339,87 @@ def writer(args, res_queue, input_file_path):
     out_base_path = os.path.join(args.exp_dir, prefix + "_" + args.result_fname)
     if args.task == "regression":
         channels = [0]
-        outfiles = [open(os.path.join(out_base_path + ".track.bedGraph"), "w")]
+        outfiles = [os.path.join(out_base_path + ".track.bedGraph")]
     elif args.task == "classification":
         channels = [1]
-        outfiles = [open(os.path.join(out_base_path + ".peaks.bedGraph"), "w")]
+        outfiles = [os.path.join(out_base_path + ".peaks.bedGraph")]
     elif args.task == "both":
         channels = [0, 1]
-        outfiles = [open(os.path.join(out_base_path + ".track.bedGraph"), "w"),
-                    open(os.path.join(out_base_path + ".peaks.bedGraph"), "w")]
+        outfiles = [os.path.join(out_base_path + ".track.bedGraph"),
+                    os.path.join(out_base_path + ".peaks.bedGraph")]
+
+    # Temp dir used to save temo files during multiprocessing.
+    temp_dir = tempfile.mkdtemp()
+    for channel in channels:
+        os.makedirs(os.path.join(temp_dir, str(channel)))
+
     while 1:
         if not res_queue.empty():
             item = res_queue.get()
             if item == 'done':
                 break
-            for channel in channels:
-                batch_bg = get_batch_bg(args, item, channel, intervals)
-                df_to_bedGraph(batch_bg, outfiles[channel])
+            keys, batch = item
+            batch_num = keys.numpy()[0]
+            #No multi processing
+            if args.num_workers == 0:
+                start = 0
+                end = batch.shape[0]
+                for channel in channels:
+                    batch_bg = save_to_bedgraph([start, end], item, channel, intervals, outfiles[channel], args.round)
+            else:
+                num_jobs = math.ceil(batch.shape[0] / args.batches_per_worker)
+                pool_size = 0
+
+                if args.num_workers == -1: # spawn pool of processes
+                    num_cpus = mp.cpu_count()
+                    pool_size = min(num_jobs, num_cpus)
+                else: # user specified # of processes
+                    pool_size = args.num_workers
+
+                pool = mp.Pool(pool_size)
+                tempfiles = [os.path.join(temp_dir, str(channel), "{0:05}".format(num+batch_num)) for channel in channels for num in range(num_jobs)]
```

Comment:
few things here - 
1. let's separate the prefix and suffix of the file name with `_` (e.g. `tempdir0_num`)
2. when doing (num + batch_num), is that sum required to be unique? Asking coz job num 0 + batch num 1 == job num 1 + batch num 0
3. let's rename this var to `temp_files` to keep it similar to other vars

Reply:
it's not a prefix, it's a subdir. `tempdir/0/(num+batch_num)`. Makes sense ?

Reply:
My bad, didn't see the `makedirs` call up top! All good.

And I'm guessing 2 wouldn't matter because in each loop all files are combined into the outfile anyway

Reply:
Can you fix the var name to be `temp_files`?

Reply:
Yes sorry, I only gave half answer. Yes, 2 wouldn't matter. I will change the file name.

Reply:
done.